### PR TITLE
Under the hood fixes

### PR DESCRIPTION
**Pull request by shibshib**
As title states, this branch is for any fixes found along the way as we are building Adam.

Adam will eventually be forked to its own repository, but in the interest of under-the-hood fixes, we will keep this as a branch for now.

### CODE REVIEW

File: adam_config.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+axon:
```
Comment (unconst) [inline adam_config.yaml]: This shouldn't be added to the main

File: bittensor/axon.py [hunk 2]
```diff
@@ -194,6 +194,7 @@ def _forward(self, request):
                 code: (:obj:`bittensor_pb2.ReturnCode, `required`)
                     return code associated with forward call i.e. Success of Timeout.
         """
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (unconst) [inline bittensor/axon.py]: this should be pulled directly from the synapse, not from this line.

ie. synapse.device?

File: bittensor/dendrites/pkm.py [hunk 3]
```diff
@@ -113,17 +114,20 @@ def _route(self, inputs, query, modality: bittensor_pb2.Modality) -> Tuple[torch
 
         # query: (torch.FloatTensor): projection of the query on to the key dimension.
         # query.shape = [batch_size, config.dendrite.key_dim]
-        query = self.projection( query ).to(self.device)
+        # On Cuda if it's available.
+        query = self.projection( query )#.to(self.device)
```
Comment (unconst) [inline bittensor/dendrites/pkm.py]: Did you mean to have this commented?

File: bittensor/dendrites/pkm.py [hunk 4]
```diff
@@ -191,37 +195,37 @@ def _route(self, inputs, query, modality: bittensor_pb2.Modality) -> Tuple[torch
 
         # nonzero_gates: (torch.FloatTensor): non-zero gating values for each example for each uid.
         # nonzero_gates.shape = [real_topk * batch_size, 1]
-        nonzero_gates = torch.gather(gates_expanded, 1, uids_index).to(self.device)
+        nonzero_gates = torch.gather(gates_expanded, 1, uids_index)#.to(self.device)
```
Comment (unconst) [inline bittensor/dendrites/pkm.py]: again here?

File: bittensor/dendrites/pkm.py [hunk 5]
```diff
@@ -191,37 +195,37 @@ def _route(self, inputs, query, modality: bittensor_pb2.Modality) -> Tuple[torch
...
 
         # flat_stitched: (torch.FloatTensor): responses multiplied by gate values.
         # flat_stitched.shape = [real_topk * batch_size, *inputs.shape[1:]] 
         flat_stitched = flat_stitched.mul(nonzero_gates)
 
         # zeros: (torch.FloatTensor): zero for combined responses. 
         # zeros.shape = [batch_size, *inputs.shape[1:]] 
-        zeros = torch.zeros(batch_size, flat_stitched.shape[1], requires_grad=True).to(self.device)
+        zeros = torch.zeros(batch_size, flat_stitched.shape[1], requires_grad=True)#.to(self.device)
```
Comment (unconst) [inline bittensor/dendrites/pkm.py]: and here?

File: bittensor/dendrites/pkm.py [hunk 6]
```diff
@@ -191,37 +195,37 @@ def _route(self, inputs, query, modality: bittensor_pb2.Modality) -> Tuple[torch
...
+        combined = zeros.to(self.device).index_add(0, batch_index, flat_stitched.float())
 
         # combined: (torch.FloatTensor): combined responses reshaped to correct dimension.
         # combined = [batch_size, sequence_dim, bittensor.__network_dim__]
         combined = combined.view(batch_size, inputs.shape[1], bittensor.__network_dim__)
 
         # indices: (torch.LongTensor): indices of uids queried during this forward call.
         # indices = [batch_size, metagraph.n]
-        indices = self.session.metagraph.uids_to_indices(filtered_uids).to(self.device)
+        indices = self.session.metagraph.uids_to_indices(filtered_uids)#.to(self.device)
```
Comment (unconst) [inline bittensor/dendrites/pkm.py]: same

File: bittensor/synapse.py [hunk 7]
```diff
@@ -99,7 +99,6 @@ def __init__(   self,
         self.config = config
         self.session = session
         self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-        self.to(self.device)
```
Comment (unconst) [inline bittensor/synapse.py:102]: Maybe makes sense to add a special function here set_synapse_device(device)

File: bittensor/synapses/gpt2.py [hunk 8]
```diff
@@ -164,7 +160,7 @@ def forward_text(self, inputs: torch.LongTensor):
                 hidden (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_len, bittensor.__network_dim__)`, `required`): 
                     Hidden layer representation produced using the local_context.
         """
-        hidden = self.forward(inputs=inputs.to(self.device), training = False, remote = False).local_hidden
+        hidden = self.forward(inputs=inputs, training = False, remote = False).local_hidden
```
Comment (unconst) [inline bittensor/synapses/gpt2.py:163]: this is the forward call from the server, did you want to take this out? Perhaps it is better to add the to.device() in the nucleus?

File: bittensor/synapses/gpt2.py [hunk 9]
```diff
@@ -241,22 +237,24 @@ def forward(self,
         # local_hidden.shape = [batch_size, sequence_len, bittensor.__network_dim__]
         local_hidden = torch.cat([encoding, local_context], dim=2)
         local_hidden = self.hidden_layer(local_hidden)
-        output.local_hidden = local_hidden
+
```
Comment (unconst) [inline bittensor/synapses/gpt2.py]: So this would mean the local_hidden is not actually returned by the function call, it will mean the forward call from a remote user fails entirely

File: bittensor/synapses/gpt2.py [hunk 10]
```diff
@@ -241,22 +237,24 @@ def forward(self,
...
         local_hidden = torch.cat([encoding, local_context], dim=2)
         local_hidden = self.hidden_layer(local_hidden)
-        output.local_hidden = local_hidden
+        
         if training:
             # local_target: projection of local_hidden onto target dimension.
             # local_target.shape = [batch_size, sequence_len, bittensor.__vocab_size__]
             local_target = self.target_layer(local_hidden)
-            output.local_target = local_target
+            #output.local_target = local_target
```
Comment (unconst) [inline bittensor/synapses/gpt2.py]: again here, why do this?

File: bittensor/synapses/gpt2.py [hunk 11]
```diff
@@ -241,22 +237,24 @@ def forward(self,
...
             # local_target_loss: MLM loss between local_target and passed targets.
             # local_target_loss.shape = [1]
             shift_logits = local_target[..., :-1, :].contiguous()
             shift_labels = inputs[..., 1:].contiguous()
             local_target_loss = self.loss_fct(
                 shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
-            output.local_target_loss = local_target_loss
             output.loss = output.loss + local_target_loss
-
+            output.local_target_loss = local_target_loss.item()
```
Comment (unconst) [inline bittensor/synapses/gpt2.py]: if we take the item it means we are not training the local_target loss which means the distillation is not actually improving the model. Did you also take out the distillation loss? cause this will mean it does not distill

File: bittensor/synapses/gpt2.py [hunk 12]
```diff
@@ -330,19 +328,20 @@ def forward_remote(self, local_context, local_hidden, inputs, pooled, encoding,
         output.weights = weights
         output.return_codes = return_codes
         output.request_sizes = sizes 
-        remote_context = remote_context.to(self.device)
+        del weights, sizes, return_codes
```
Comment (unconst) [inline bittensor/synapses/gpt2.py]: Is this required?

File: bittensor/synapses/gpt2.py [hunk 13]
```diff
@@ -330,19 +328,20 @@ def forward_remote(self, local_context, local_hidden, inputs, pooled, encoding,
...
 
         # distillation_loss: distillation loss between local_context and remote_context
         # distillation_loss.shape = [1]
         distillation_loss = F.mse_loss(local_context, remote_context.detach())
-        output.distillation_loss = distillation_loss
         output.loss = output.loss + distillation_loss
+        output.distillation_loss = distillation_loss.item()
```
Comment (unconst) [inline bittensor/synapses/gpt2.py]: again here, this will mean the model doesn't learn to distill

File: bittensor/utils/logging.py [hunk 14]
```diff
@@ -225,17 +225,17 @@ def log_outputs(history: List[bittensor.synapse.SynapseOutput]):
             row.append('')
 
         if output.local_target_loss != None:
-            row.append(output.local_target_loss.item())
+            row.append(output.local_target_loss)
```
Comment (unconst) [inline bittensor/utils/logging.py]: seems really hacky?

File: neurons/adam.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    def train(epoch: int, global_step: int):
+        # ----- Init training state --- 
+        model.train() 
+        session.metagraph.sync() # Sync with the chain.
+        row_weights = session.metagraph.W[ 0, :] # My weights on the chain-state (zeros initially).
+        history = []
+        local_step = 0
+        local_epochs = 10
+        output = None
+        net = torch.nn.DataParallel(model)
```
Comment (unconst) [inline neurons/adam.py]: What does this do?

File: neurons/adam.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+    if config.neuron.record_log:
+        current_time = time.strftime("%H_%M_%S", time.localtime())
+        logger.add("{}_{}.log".format(config.neuron.name, current_time),format="{time:YYYY-MM-DD at HH:mm:ss} | {level} | {message}")
+
+    # ---- Train Epoch ----
+    def train(epoch: int, global_step: int):
+        # ----- Init training state --- 
+        model.train() 
+        session.metagraph.sync() # Sync with the chain.
+        row_weights = session.metagraph.W[ 0, :] # My weights on the chain-state (zeros initially).
```
Comment (unconst) [inline neurons/adam.py]: This will fail if there are no weights, can we add a check

File: adam_config.yaml (hunk 1 above)
Comment (shibshib) [reply adam_config.yaml]: This file is taken out completely now, not needed anymore as it's moved to Adam repo.

File: bittensor/axon.py (hunk 2 above)
Comment (shibshib) [reply bittensor/axon.py]: Fixed

File: bittensor/dendrites/pkm.py [hunk 17]
```diff
@@ -82,9 +82,10 @@ def _route(self, inputs, query, modality: bittensor_pb2.Modality) -> Tuple[torch
                     dendrite call return codes.
         """
         # For ease of use.
-        inputs = inputs.to(self.device)
+        #inputs = inputs.to(self.device)
```
Comment (unconst) [inline bittensor/dendrites/pkm.py]: if the line isn't required I would just take it out

File: bittensor/nucleus.py [hunk 18]
```diff
@@ -23,6 +23,7 @@ def __init__(self, config):
         self._config = config
         self._forward_pool = ThreadPoolExecutor(maxsize = self._config.nucleus.queue_maxsize, max_workers=self._config.nucleus.max_workers)
         self._backward_pool = ThreadPoolExecutor(maxsize = self._config.nucleus.queue_maxsize, max_workers=self._config.nucleus.max_workers)
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (unconst) [inline bittensor/nucleus.py]: this should be looking at the synapse to see if the synapse.device is set

**Review by unconst — APPROVED** [review]
